app/models/media_attachment.rb

```diff
@@ -106,13 +106,18 @@ def set_shortcode
   end
 
   def set_type_and_extension
-    if file.blank?
-      self.type = :unknown
-    else
-      self.type = VIDEO_MIME_TYPES.include?(file_content_type) ? :video : :image
-      extension = Paperclip::Interpolations.content_type_extension(file, :original)
-      basename  = Paperclip::Interpolations.basename(file, :original)
-      file.instance_write :file_name, [basename, extension].delete_if(&:empty?).join('.')
-    end
+    self.type = VIDEO_MIME_TYPES.include?(file_content_type) ? :video : :image
+    extension = appropriate_extension
+    basename  = Paperclip::Interpolations.basename(file, :original)
+    file.instance_write :file_name, [basename, extension].delete_if(&:empty?).join('.')
+  end
+
+  def appropriate_extension
+    mime_type = MIME::Types[file.content_type]
+
+    extensions_for_mime_type = mime_type.empty? ? [] : mime_type.first.extensions
+    original_extension       = Paperclip::Interpolations.extension(file, :original)
+
+    extensions_for_mime_type.include?(original_extension) ? original_extension : extensions_for_mime_type.first
```

Comment:
Basically we have to fight paperclip here a bit because what we're doing is more complicated than the library is meant to handle. Here is what we *want* to do: if it's an animated gif, convert it to video. If not, leave as is. But at the same time, we want to modify the *original* files, because the whole point is conserving space. However, paperclip doesn't handle modifications of the original file very well.

What you see here is the result of a lot of troubleshooting with `binding.pry` and reading paperclip's source code. It's not pretty code, and it's not very logical, but it works.